Hw21 cache

ihavedone opened the pull request.

ihavedone committed:
init

ihavedone committed:
hibernate markup & using

ihavedone committed:
add account (this class doesn't take a part in the test and hibernate, it is just a stub for the future).
fix dual link (user <--> phones, user <--> address)
add some tests

ihavedone committed:
init (code from the previous work with hibernate)

ihavedone committed:
add cache interfaces

ihavedone committed:
myCache implementation

ihavedone committed:
cache implementation and checking

ihavedone committed:
evidence

petrelevich commented on settings.gradle:
```diff
@@ -7,4 +7,6 @@ include 'hw09-aop'
 include 'hw15-serialization'
-include 'hw17-jdbc'
+include 'hw19-hibernate'
+include 'hw19-hibernate'
```
тут include 'hw21-cache'
должно быть.

petrelevich commented on hw21-cache/src/main/java/ru/otus/merets/cachehw/MyCache.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public void put(K key, V value) {
+        listeners.forEach(p->p.notify(key,value, "before put"));
```
не забывайте про форамтирование кода (пробелы).
В Idea это alt-ctrl-l

petrelevich commented on hw21-cache/src/main/java/ru/otus/merets/cachehw/MyCache.java:
```diff
@@ -0,0 +1,78 @@
+    public void put(K key, V value) {
+        listeners.forEach(p->p.notify(key,value, "before put"));
+        cache.put(key, value);
+        listeners.forEach(p->p.notify(key,value, "after put"));
```
p.notify можно в try-catch вызывать, чтобы случайное исключение цепочку вызовов не нарушило.

petrelevich commented on hw21-cache/src/main/java/ru/otus/merets/core/service/DbServiceUserImpl.java:
```diff
@@ -0,0 +1,91 @@
+import ru.otus.merets.core.dao.UserDao;
+import ru.otus.merets.core.model.User;
+
+import javax.swing.text.html.Option;
```
Удаляйте ненужные импотрты, особенно, которые по ошибке добавлены.

petrelevich commented on hw21-cache/src/main/java/ru/otus/merets/core/service/DbServiceUserImpl.java:
```diff
@@ -0,0 +1,91 @@
+    public Optional<User> getUser(long id) {
+        Optional<User> userOptional = Optional.ofNullable(cache.get(String.valueOf(id)));
+
+        if (!userOptional.isPresent()) {
```
можно так:
if (userOptional.isEmpty())

petrelevich commented on hw21-cache/src/main/java/ru/otus/merets/core/service/DbServiceUserImpl.java:
```diff
@@ -0,0 +1,91 @@
+        return userOptional;
+    }
+
+    public Optional<User> getUserNoCache(long id) {
```
Этот метод очень сильно портит api.
кеш можно передавать как параметр конструктора.
Если передан, то использовать.
Тогда и этот метод не потребудется.

ihavedone committed:
cache fix:
- API (remove nocache method from DbServiceUserImpl);
- add try\catch to listeners notifications
